### PR TITLE
[1.8] Automated cherry pick of #58720 #57326 #60342

### PR DESCRIPTION
**Pull request by joelsmith**
Cherry pick of #58720 #57326 #60342 on release-1.8.

#58720: Ensure that the runtime mounts RO volumes read-only
#57326: Update Dashboard version to v1.8.3
#60342: Fix nested volume mounts for read-only API data volumes

Fixes https://github.com/kubernetes/kubernetes/issues/60814 for 1.8

Note for reviewers:
The Dashboard update is necessary because previous versions of the Dashboard attempted to write to the secret volume on startup, and so the read-only secret volume change prevented the Dashboard from starting. 

```release-note
Changes secret, configMap, downwardAPI and projected volumes to mount read-only, instead of allowing applications to write data and then reverting it automatically. Until version 1.11, setting the feature gate ReadOnlyAPIDataVolumes=false will preserve the old behavior. Updates dashboard version to v1.8.3 to avoid writing to read-only locations.
```

### CODE REVIEW

File: cluster/addons/dashboard/dashboard-controller.yaml [hunk 1]
```diff
@@ -56,8 +56,7 @@ spec:
           timeoutSeconds: 30
       volumes:
       - name: kubernetes-dashboard-certs
-        secret:
-          secretName: kubernetes-dashboard-certs
+        emptyDir: {}
```
Comment (rata) [inline cluster/addons/dashboard/dashboard-controller.yaml]: I see there was an issue opened on the dashboard discussed before, https://github.com/kubernetes/kubernetes/pull/58720/files#r166033248, but it's better to back port like this, right?

File: pkg/volume/configmap/configmap.go [hunk 2]
```diff
@@ -194,6 +194,9 @@ func (b *configMapVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
 	if err := wrapped.SetUpAt(dir, fsGroup); err != nil {
 		return err
 	}
+	if err := volumeutil.MakeNestedMountpoints(b.volName, dir, b.pod); err != nil {
```
Comment (rata) [inline pkg/volume/configmap/configmap.go]: Seems disturbing to use `b` here and `*b` here: https://github.com/kubernetes/kubernetes/pull/60515/commits/25f34665f8127ec3466c06e733da423772a17f48#diff-a3b28c0b2e7de9ad275e0880d0764c8dR187

But it's like the original PR, so perfect :)

File: cluster/addons/dashboard/dashboard-controller.yaml (hunk 1 above)
Comment (rata) [reply cluster/addons/dashboard/dashboard-controller.yaml]: Oh, now I see the other commit using the new dashboard. Dismiss this, please :)

**Review by rata — APPROVED** [review]
LGTM
Thanks a lot!

I reviewed commit by commit, for more easy inspect what belongs to which PR, and some lines were later modified in another commit. Sorry for the comments that were stale because the new commit fixed it :)

**Review by feiskyer — COMMENTED** [review]
hack/verify-bazel.sh failed, need to run hack/update-bazel.sh

**Comment by joelsmith** [comment]
/retest
Now that #60523 is merged, maybe it'll pass

**Comment by joelsmith** [comment]
I didn't pay enough attention on the backport of the dashboard update. One of its changes prevented the dashboard from starting on 1.8:
```patch
-apiVersion: apps/v1beta2
+apiVersion: apps/v1
```
So I've reverted that change (and squashed the revert into the commit)

Also, the failing test seems to have passed all the test cases but had some problem during the teardown of the test cluster.
/retest

**Comment by BenTheElder** [comment]
Apologies, we had a small migration that required restarting the queue (`pull-kubernetes-unit` -> `pull-kubernetes-integration`), we've been waiting for a quiet enough day. See: https://github.com/kubernetes/test-infra/pull/7057#issuecomment-369384392

Getting this back on track for merge:
/test pull-kubernetes-integration

**Comment by BenTheElder** [comment]
PR is now back in the queue (and next)

**Pull request merged**